### PR TITLE
Address gh-11515, improve in-practice runtime to insert clean boolean data, and improve source code readability

### PR DESCRIPTION
**Pull request by GNU-Linuxer**
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory.

If you're making a change to documentation, do **not** modify a `.html` file directly. Instead find the corresponding `.pug` file or test case in the `test/docs` directory.

**Summary**

Solves gh-11515 (https://github.com/Automattic/mongoose/issues/11515) pluralization concern by rebasing from pluralize.js module (MIT license, [commit](https://github.com/plurals/pluralize/blob/36f03cd2d573fa6d23e12e1529fa4627e2af74b4/pluralize.js)), along with test suite.

Improved in-practice runtime to insert clean boolean data by skipping Set creations if the input is a boolean-typed value.

Extracted similar operations in `Schema._clone()` function to declutter the code.

`test/model.create.test.js`: renamed variables

`lib/helpers/clone.js`: removed confusing comment

`lib/schema.js`: fixed variable name and changed function name for better readability

**Examples**

Added tests for the improved pluralization (gh-11515) code.

Added tests that validates casting values to boolean.

### CODE REVIEW

File: lib/helpers/pluralize.js
Comment (AbdelrahmanHafez) [inline lib/helpers/pluralize.js:9]: Please remove the `// David Xie: ...` comments that mention the commits, they don't serve much value.

File: lib/helpers/clone.js
Comment (AbdelrahmanHafez) [inline lib/helpers/clone.js:115]: This is used to avoid automatically generating docs for `cloneObject` since it's meant to be used internally and not part of the public API.

File: lib/schema.js
Comment (AbdelrahmanHafez) [inline lib/schema.js:314]: Using different terminology(`schemaDeepCopy` vs `schema.clone`) can be a bit confusing in the long run.
Perhaps `function cloneSchema() {...}` would be more consistent?

File: lib/schema.js
Comment (AbdelrahmanHafez) [inline lib/schema.js:351]: [Please don't commit commented out code](https://kentcdodds.com/blog/please-dont-commit-commented-out-code).

File: test/model.create.test.js
Comment (AbdelrahmanHafez) [inline test/model.create.test.js:76]: I would expect a variable with a verb to be a function, perhaps `promise`/`value`?

File: test/helpers/pluralize.test.js
Comment (AbdelrahmanHafez) [inline test/helpers/pluralize.test.js:19]: We don't need to have a `done();` here anymore, since mocha now runs synchronous tests to the end and will throw an error if the assertion fails.

File: lib/error/serverSelection.js
Comment (AbdelrahmanHafez) [inline lib/error/serverSelection.js:37]: I agree that `includes` provides better readability, but after we merge this, the comment won't mean much in the new context.

Nitpick: I think we can make it even more explicit to `err.message.includes('bad auth') === false` over `!err.message.includes('bad auth')` because it's much easier to not notice the `!` than the ` === false`, what do you think?

**Comment by AbdelrahmanHafez** [comment]
Welcome, @GNU-Linuxer 👋

I've posted a few comments, also some tests are failing due to the proposed changes, can you fix those?
Overall, those feel like heavy changes and potentially breaking. I'd rather ship those on v7.0 maybe.

What do you think? @GNU-Linuxer @vkarpov15

File: lib/helpers/pluralize.js
Comment (vkarpov15) [inline lib/helpers/pluralize.js:14]: We unfortunately cannot merge this change until Mongoose 7, because of all of the backwards breaking changes here.

File: lib/helpers/pluralize.js
Comment (vkarpov15) [inline lib/helpers/pluralize.js:10]: Given that Mongoose is Apache 2 licensed, based on my understanding we can't have part of the code being MIT... Is there any way we can simply just use a fixed version of pluralize instead?

File: lib/schema.js
Comment (vkarpov15) [reply lib/schema.js:314]: I agree, this change doesn't seem necessary or helpful. Maybe `schemaClone()`, but I'd prefer to make this `function clone()`

File: lib/schema.js
Comment (vkarpov15) [inline lib/schema.js:310]: Why is this removed? We still want to keep this content for the docs.

File: test/model.create.test.js
Comment (vkarpov15) [inline test/model.create.test.js:153]: `callback` refers to a value of type promise here, that's confusing

**Review by vkarpov15 — CHANGES_REQUESTED** [review]
In the future, can you please split up related changes into separate PRs? This PR is a hodge podge of unrelated changes; some very minor, some significant, and some heavily backwards breaking. I'd love to merge some of these changes for immediate release, but with the pluralize changes and the unexpected removal of `Schema.prototype.clone()` docs there's no way for us to easily do so.

**Comment by vkarpov15** [comment]
I'm going to close this PR. Not really much we can do with this. Feel free to re-open in smaller PRs with similar changes grouped together.

**Pull request closed without merging**
